File: Recursion/permutation.js
function allPermutation(nums) {
    const permutation = [];
    function helper(nums, i) {
        if (i === nums.length - 1) {
            permutation.push(nums.slice());
            return;
        }
        for (let j = i; j < nums.length; j++) {
            // swap i and j 
            [nums[i], nums[j]] = [nums[j], nums[i]];
            //recursive
            helper(nums, i + 1);
            // swap i and j
            [nums[i], nums[j]] = [nums[j], nums[i]];
        }
    }
    helper(nums, 0);
    return permutation;
}

console.log(allPermutation([1, 2, 3]));

File: Recursion/permutation.py
def all_permutation(nums):
    permutations = []
    if len(nums) == 0: return [[]]
    def helper(nums,i):
        if i == len(nums)-1:
            permutations.append(nums.copy())
            return
        for j in range(i,len(nums)):
            nums[i],nums[j] = nums[j], nums[i]
            helper(nums, i+1)
            nums[i],nums[j] = nums[j], nums[i]
            
    helper(nums,0)
    return permutations 

print(all_permutation([1,2,3]))
    